File: src/Aspire.Dashboard/Components/ResourcesGridColumns/UrlsColumnDisplay.razor.js
// Adaptive overflow measurement for the resource URLs column. Renders inline URL items and, using a
// ResizeObserver, decides how many fit as the column width changes; the first item is always kept.
// Overflowed items are hidden inline (they remain in the popover the component renders) and the
// component is told the visible count so it can update the "+N" button and popover contents.
// Colocated module: no inline script, so it's CSP-safe. The observer is disconnected in dispose().

const registrations = new WeakMap();

// Space (px) reserved for the trailing "+N" button when there is overflow. A little larger than the
// button so a partially-fitting item doesn't sit flush against it.
const MORE_BUTTON_RESERVE = 52;

export function initialize(container, dotNetReference) {
    const overflow = container.querySelector("[data-url-overflow]");
    if (!overflow) {
        return;
    }

    let lastVisible = -1;

    const measure = () => {
        const items = Array.from(overflow.querySelectorAll("[data-url-item]"));
        if (items.length === 0) {
            return;
        }

        // Reveal every item first so we can read its intrinsic width, including items hidden on the
        // previous pass. This runs inside the ResizeObserver/interop callback (before paint), so the
        // reveal-measure-hide cycle below doesn't flash on screen.
        for (const item of items) {
            item.style.display = "";
        }

        const available = overflow.clientWidth;

        let total = 0;
        for (const item of items) {
            total += item.getBoundingClientRect().width;
        }

        let visible;
        if (total <= available) {
            // Everything fits: no overflow, so no button space needs reserving.
            visible = items.length;
        } else {
            // Fit as many as possible while leaving room for the "+N" button. The first item is
            // always visible (it truncates with an ellipsis if it's wider than the column).
            const budget = available - MORE_BUTTON_RESERVE;
            let used = 0;
            visible = 0;
            for (let i = 0; i < items.length; i++) {
                const width = items[i].getBoundingClientRect().width;
                if (i === 0) {
                    used += width;
                    visible = 1;
                    continue;
                }
                if (used + width <= budget) {
                    used += width;
                    visible++;
                } else {
                    break;
                }
            }
        }

        // Apply the computed visibility synchronously so there's no flash of overflowing items.
        for (let i = 0; i < items.length; i++) {
            items[i].style.display = i < visible ? "" : "none";
        }

        if (visible !== lastVisible) {
            lastVisible = visible;
            dotNetReference.invokeMethodAsync("SetVisibleCountAsync", visible);
        }
    };

    const observer = new ResizeObserver(() => measure());
    observer.observe(overflow);

    registrations.set(container, { observer, measure });

    // Initial measure so the layout is correct before the first resize event.
    measure();
}

export function measure(container) {
    const registration = registrations.get(container);
    if (registration) {
        registration.measure();
    }
}

export function dispose(container) {
    const registration = registrations.get(container);
    if (registration) {
        registration.observer.disconnect();
        registrations.delete(container);
    }
}

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/UrlsColumnDisplay.razor.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using Aspire.Dashboard.Model;
using Aspire.Dashboard.Resources;
using Microsoft.AspNetCore.Components;
using Microsoft.Extensions.Localization;
using Microsoft.JSInterop;

namespace Aspire.Dashboard.Components;

public partial class UrlsColumnDisplay : IAsyncDisposable
{
    // Safety cap on how many URL elements are ever rendered inline. Rendering hundreds of URL DOM
    // elements triggers a forced synchronous reflow long enough to drop the SignalR connection, so
    // anything past the cap is only ever shown in the overflow popover.
    private const int MaxRenderedUrls = 20;

    internal static string GetTooltipText(DisplayedUrl displayedUrl)
    {
        return displayedUrl.Url ?? displayedUrl.OriginalUrlString;
    }

    [Parameter, EditorRequired]
    public required ResourceViewModel Resource { get; set; }

    [Parameter, EditorRequired]
    public required bool HasMultipleReplicas { get; set; }

    [Parameter, EditorRequired]
    public required IList<DisplayedUrl> DisplayedUrls { get; set; }

    [Parameter]
    public string? AdditionalMessage { get; set; }

    [Inject]
    public required IStringLocalizer<Columns> Loc { get; init; }

    [Inject]
    public required IJSRuntime JS { get; init; }

    private bool _popoverVisible;

    // Stable id for the "+N" overflow button so the popover can anchor to it.
    private readonly string _moreButtonId = $"urls-more-{Guid.NewGuid():N}";

    private ElementReference _container;
    private IJSObjectReference? _module;
    private DotNetObjectReference<UrlsColumnDisplay>? _selfReference;

    // Number of the (up to MaxRenderedUrls) rendered items shown inline. Until the JS measurer runs
    // we optimistically assume everything fits; the measurer collapses overflow into the popover.
    private int _visibleCount = MaxRenderedUrls;

    // Signature of the currently rendered URL set, used to re-measure only when the data changed
    // (a ResizeObserver in the module already handles width changes on its own).
    private string? _measuredSignature;

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        // Only the overflow layout needs measuring; a single (or no) URL is laid out by the browser.
        if (DisplayedUrls.Count <= 1)
        {
            return;
        }

        if (firstRender)
        {
            _module = await JS.InvokeAsync<IJSObjectReference>("import", "./Components/ResourcesGridColumns/UrlsColumnDisplay.razor.js");
            _selfReference = DotNetObjectReference.Create(this);
            _measuredSignature = ComputeSignature();
            await _module.InvokeVoidAsync("initialize", _container, _selfReference);
        }
        else if (_module is not null)
        {
            // Re-measure when the rendered URLs changed. The ResizeObserver won't fire for a
            // content-only change that leaves the column the same width.
            var signature = ComputeSignature();
            if (signature != _measuredSignature)
            {
                _measuredSignature = signature;
                await _module.InvokeVoidAsync("measure", _container);
            }
        }
    }

    /// <summary>
    /// Called by the resize measurer with the number of rendered items that fit inline. The first
    /// item is always kept, and the count is clamped to what we actually rendered.
    /// </summary>
    [JSInvokable]
    public Task SetVisibleCountAsync(int visibleCount)
    {
        var renderedCount = Math.Min(DisplayedUrls.Count, MaxRenderedUrls);
        var clamped = Math.Clamp(visibleCount, 1, Math.Max(1, renderedCount));
        if (clamped != _visibleCount)
        {
            _visibleCount = clamped;
            StateHasChanged();
        }

        return Task.CompletedTask;
    }

    // Cheap identity of the rendered subset: count plus each rendered item's original URL. Small
    // because it only ever covers up to MaxRenderedUrls items.
    private string ComputeSignature()
    {
        return $"{DisplayedUrls.Count}\u0001{string.Join('\u0001', DisplayedUrls.Take(MaxRenderedUrls).Select(u => u.OriginalUrlString))}";
    }

    public async ValueTask DisposeAsync()
    {
        try
        {
            if (_module is not null)
            {
                await _module.InvokeVoidAsync("dispose", _container);
                await _module.DisposeAsync();
            }
        }
        catch (JSDisconnectedException)
        {
            // Circuit already gone; nothing to clean up.
        }

        _selfReference?.Dispose();
    }
}
